Use enum field for HotThreads report type

grcevski opened the pull request:
HotThreads report type field is currently declared as a String, but internally it is used as an enum. This PR changes the report type to enum and does a small refactor to encapsulate the getter function with the accumulator class.

I also extended the HotThreads unit tests to test some of the newly refactored code.

Note: This change introduces slight behaviour difference. Previously when the HotThreads API was called with an incorrect report type, e.g. /_nodes/hot_threads?type=gpu we'd crash with InvalidArgumentException, but after argument validation, therefore producing an empty 400 response. After this change we get a 400 with response: "reason": "type not supported [gpu]"

Commit:
Switch HotThreads request type to enum

HotThreads report request type was declared as String, but used as an
enum where only few values were allowed. This PR changes the type to
enum allowing malformed requests to correctly report the error.

This PR also introduces a small refactor to make the report getter
function part of the accumulator class.

Commit:
Remove automatic variable inference from tests

Commit:
Fix missed type to String conversion

grcevski commented:
This PR is the last bullet point of the broader, and now closed PR https://github.com/elastic/elasticsearch/pull/77392

elasticmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

grcevski commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into enhancement/hotthreads_enum_type

rjernst commented on server/src/main/java/org/elasticsearch/action/admin/cluster/node/hotthreads/NodesHotThreadsRequest.java:
```diff
@@ -29,7 +30,7 @@ public NodesHotThreadsRequest(StreamInput in) throws IOException {
         super(in);
         threads = in.readInt();
         ignoreIdleThreads = in.readBoolean();
-        type = in.readString();
```
This isn't going to be wire compatible, as readEnum uses ordinals for serialization, not the string name. We will need to version this change, using `in.getVersion().onOrAfter(Version.V_7_16_0)` as a condition for the readEnum call (assuming we will backport to 7.x).

Another thing to consider is that because readEnum uses ordinals, any change in the ordering of the ReportType values could cause serialization changes. It might be beneficial to (separately) look at making readEnum use strings internally, but then use `valueOf` to get the enum value from string.

grcevski replied:
Thanks for this find! I didn't think of the mix cluster case and I thought handling it at the REST layer will be enough. I'll revert the request type to String to be backwards compatible and use the .of helper to convert to the enum.

rjernst commented on server/src/main/java/org/elasticsearch/action/admin/cluster/node/hotthreads/NodesHotThreadsRequest.java:
```diff
@@ -92,7 +93,7 @@ public void writeTo(StreamOutput out) throws IOException {
         out.writeInt(threads);
         out.writeBoolean(ignoreIdleThreads);
-        out.writeString(type);
+        out.writeEnum(type);
```
You'll need a similar version conditional here to match the read conditional.

grcevski replied:
:+1 I'll revert back and make it compatible.

rjernst commented on server/src/main/java/org/elasticsearch/monitor/jvm/HotThreads.java:
```diff
@@ -31,15 +31,42 @@
+            return type;
+        }
+
+        public static ReportType of(String type) {
```
Don't all enums have `valueOf` that we can use?

grcevski replied:
The only reason I made a custom 'valueOf' is because I wanted to throw a nicer message in case the enum type doesn't exist. Enum.valueOf will say something like 'no enum constant...' with internal information about the class, which will be confusing to the users. It's a static method, so we can't override it.

rjernst commented on server/src/test/java/org/elasticsearch/monitor/jvm/HotThreadsTests.java:
```diff
@@ -307,13 +307,31 @@ public void testEnsureInnerDetectSkipsCurrentThread() throws Exception {
             .ignoreIdleThreads(false);
 
-        var innerResult = hotThreads.innerDetect(mockedMXBean, mockCurrentThreadId);
+        String innerResult = hotThreads.innerDetect(mockedMXBean, mockCurrentThreadId);
```
Are these type changes to make the backport easier?

grcevski replied:
Yeah, I started using inferred types, but makes extending the tests hard for 7.x and I plan to add few more unit tests.

rjernst reviewed (changes requested):
Looks pretty good. The primary thing still needed is backcompat in the serialization.

Commit:
Fix backwards compatibility

My previous change would break compatibility with older versions in
a mixed cluster. We resolve this by reading and writing a string value
instead of an enum directly.

I also added a test to ensure the serialized values are compatible with
older versions for further changes to HotThreads.

Commit:
Fix style issues

Commit:
Merge remote-tracking branch 'grcevski/enhancement/hotthreads_enum_type' into enhancement/hotthreads_enum_type

Commit:
Fix comment

Commit:
Merge branch 'master' into enhancement/hotthreads_enum_type

grcevski commented:
@rjernst I made the test backwards compatible and wrote a test to ensure it is bwc even though we use an Enum under the covers now.

rjernst commented on server/src/test/java/org/elasticsearch/action/admin/cluster/hotthreads/NodesHotThreadsRequestTests.java:
```diff
@@ -0,0 +1,84 @@
+        outHelper.writeTo(out);
+        // Write manually some values that differ from the defaults
+        // in NodesHotThreadsRequest
+        out.writeInt(4); // threads
```
Instead of manually creating a stream, you can set the version on the stream and check round trip. This is more robust as future changes may be made to the format. There are helpers to get a random version constant to test with, see `VersionUtils`.

grcevski replied:
Thanks for this! I'll look into this and make the change with my follow-up PR in the same area.

rjernst reviewed (approved):
This looks good. My comment about serialization testing could be addressed in a followup.
